packages/storage/src/database/global/queue.ts

```diff
@@ -48,65 +48,42 @@ export class ModelQueueTask extends ModelGeneral<WithoutId<TQueueRecord>> {
     );
   }
 
-  public async getLatestQueuedTaskByDatabase(
+  /** Acquires the next qualified task from the queue. The task is immediately
+   * marked as 'Proving', which prevents other workers from processing it. The
+   * calling process must ensure complete task processing and appropriate
+   * status updates (e.g., to 'Failed' or 'Proved'). A task left in 'Proving'
+   * status will remain stuck in the queue indefinitely until retry mechanism
+   * is implemented (e.g. a processing timeout). */
+  public async acquireNextTaskInQueue(
     session?: ClientSession
   ): Promise<TQueueRecord | null> {
-    if (!this.collection) {
-      throw new Error('TaskQueue is not connected to the database.');
-    }
-
-    const executingDatabases = await this.collection
-      .aggregate(
-        [
-          {
-            $match: {
+    // The intent is to:
+    // - Not allow parallel processing of tasks from the same database
+    // - Take the oldest queued task from each available database
+    // TODO: Possible edge cases:
+    // - A task stuck at Proving status will block all other tasks from the
+    // same database. Maybe consider a timeout?
+    return this.collection.findOneAndUpdate(
+      {
+        status: EProofStatusDocument.Queued,
+        databaseName: {
+          $nin: await this.collection.distinct(
+            'databaseName',
+            {
               status: EProofStatusDocument.Proving,
```

Comment:
how about Proved?

Reply:
I think we should filter the Proved also

Reply:
no, this check is to make sure that:
```
    // - Not allow parallel processing of tasks from the same database
```

that is, if a database has any `Proving` task, we exclude them from qualified next task to pop